### PR TITLE
New App: Slippirank

### PR DESCRIPTION
**Pull request by noahpodgurski**
# Description
Shows current rank from SSBM Slippi profile

### Configuration

`User Code` users Slippi code

`Show Elo` bool to show elo number

`Show rank` bool to show rank name

Images are saved and loaded with base 64

# Copilot

### Summary
📄🐍🎮

This pull request adds a new applet called SlippiRank, which shows the online rank of a Super Smash Bros. Melee player from the Slippi service. It includes a `manifest.yaml` file with the applet metadata, and a `slippirank.star` file with the applet logic and schema. The applet uses the Slippi GraphQL API to fetch the rank data, and renders it on the Tidbyt device.

> _`SlippiRank` applet_
> _Shows Melee rank from online_
> _A winter project_

### Walkthrough
*  Add a new applet for showing Slippi online rank from Super Smash Bros. Melee ([link](https://github.com/tidbyt/community/pull/1504/files?diff=unified&w=0#diff-45b9014442e609fd1eefeb20766bedec1a3c056796da29978a51638a7b99868bR1-R8), [link](https://github.com/tidbyt/community/pull/1504/files?diff=unified&w=0#diff-b807c53b5a86aba84811af1084865de5ac75c0e7ac08b66f3bfc9129e6cd4481R1-R257))
  - Create a manifest file `manifest.yaml` with the applet metadata ([link](https://github.com/tidbyt/community/pull/1504/files?diff=unified&w=0#diff-45b9014442e609fd1eefeb20766bedec1a3c056796da29978a51638a7b99868bR1-R8))
  - Write the main code in `slippirank.star` using Starlark modules and functions ([link](https://github.com/tidbyt/community/pull/1504/files?diff=unified&w=0#diff-b807c53b5a86aba84811af1084865de5ac75c0e7ac08b66f3bfc9129e6cd4481R1-R257))
    * Import modules for caching, JSON, HTTP, rendering, and schema ([link](https://github.com/tidbyt/community/pull/1504/files?diff=unified&w=0#diff-b807c53b5a86aba84811af1084865de5ac75c0e7ac08b66f3bfc9129e6cd4481R1-R257))
    * Define constants for rank names, images, ranges, and API URL ([link](https://github.com/tidbyt/community/pull/1504/files?diff=unified&w=0#diff-b807c53b5a86aba84811af1084865de5ac75c0e7ac08b66f3bfc9129e6cd4481R1-R257))
    * Define functions for getting rank name, parsing user code, requesting rank data, and caching data ([link](https://github.com/tidbyt/community/pull/1504/files?diff=unified&w=0#diff-b807c53b5a86aba84811af1084865de5ac75c0e7ac08b66f3bfc9129e6cd4481R1-R257))
    * Define the main function that takes user code and config options and returns a render.Root object ([link](https://github.com/tidbyt/community/pull/1504/files?diff=unified&w=0#diff-b807c53b5a86aba84811af1084865de5ac75c0e7ac08b66f3bfc9129e6cd4481R1-R257))
    * Define the schema function that specifies the config fields and icons ([link](https://github.com/tidbyt/community/pull/1504/files?diff=unified&w=0#diff-b807c53b5a86aba84811af1084865de5ac75c0e7ac08b66f3bfc9129e6cd4481R1-R257))

### CODE REVIEW

**Comment by noahpodgurski** [comment]
slippirank

**Comment by noahpodgurski** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by noahpodgurski** [comment]
recheck

**Comment by matslina** [comment]
Hey. This is currently not (easily) reviewable in Github due to the large file size of the starlark script itself. I think that's mostly due to the embedded graphics. You should be able to reduce the size by a lot by just scaling down the graphics to roughly the size you'll be displaying them at in the app.

**Comment by noahpodgurski** [comment]
> Hey. This is currently not (easily) reviewable in Github due to the large file size of the starlark script itself. I think that's mostly due to the embedded graphics. You should be able to reduce the size by a lot by just scaling down the graphics to roughly the size you'll be displaying them at in the app.

Updated, but it still seems it's too much for github, sorry

File: apps/slippirank/slippirank.star [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+    userCode = userCode.upper()
+    userCodeDashIndex = getUserCodeDashIndex(userCode)
+    if userCodeDashIndex == -1:
+        fail("Invalid user code")
+
+    # print(userCode)
+
+    userCodeHash = userCode[:userCodeDashIndex] + "#" + userCode[userCodeDashIndex + 1:]
+    rankedData = cache.get("rankedData")
```
Comment (matslina) [inline apps/slippirank/slippirank.star:187]: The cache module is scoped to the app, but not to individual users of the app, or individual installations of the app. In this case, the `rankedData` key will be used by all users, regardless of the configured `userCode`.

The simplest, and best fix would likely be to switch to using the built in cache functionality of the http module. Simply pass `ttl_seconds` to `http.post()` and the response will be cached accordingly.

**Review by matslina — CHANGES_REQUESTED** [review]

**Review by matslina — APPROVED** [review]

**Comment by matslina** [comment]
Excellent. Thanks for the changes!

**Pull request merged**